unit_scaling/optim.py

```diff
@@ -0,0 +1,227 @@
+    models. For example::
+
+        torch.optim.Adam(uu.optim.scaled_parameters(
+            model.parameters(), uu.optim.adam_lr_scale_func, lr=1.0
+        ))
+
+    Args:
+        params (ParamsT): an iterable of parameters of parameter groups, as passed to
+            a torch optimizer.
+        lr_scale_func (Callable): gets the optimizer-appropriate learning rate scale,
+            based on a parameter tagged with `mup_type` and `mup_scaling_depth`. For
+            example, :func:`lr_scale_func_sgd`.
+        lr (float, optional): global learning rate (overridden by groups).
+        weight_decay (float, optional): weight decay value (overridden by groups).
+        independent_weight_decay (bool, optional): enable lr-independent weight decay,
+            which performs an update per-step that does not depend on lr.
+        allow_non_unit_scaling_params (bool, optional): by default, this method fails
+            if passed any regular non-unit-scaled params; set to `True` to disable this
+            check.
+
+    Returns:
+        ParamsT: for passing on to the optimizer.
+    """
+
+    result = []
+    for entry in params:
+        group = dict(params=[entry]) if isinstance(entry, Tensor) else entry.copy()
+        group.setdefault("lr", lr)
+        group.setdefault("weight_decay", weight_decay)
+        if group["lr"] is None:
+            raise ValueError(
+                "scaled_params() requires lr to be provided,"
+                " unless passing parameter groups which already have an lr"
+            )
+        for param in group["params"]:
+            # Careful not to overwrite `lr` or `weight_decay`
+            param_lr = group["lr"]
+            if has_parameter_data(param):
+                if isinstance(param_lr, Tensor):
+                    param_lr = param_lr.clone()
+                param_lr *= lr_scale_func(param)
+            elif not allow_non_unit_scaling_params:
+                raise ValueError(
+                    "Non-unit-scaling parameter (no mup_type),"
+                    f" shape {tuple(param.shape)}"
+                )
+            param_weight_decay = group["weight_decay"]
+            if independent_weight_decay:
+                # Note: only independent of peak LR, not of schedule
+                param_weight_decay /= float(param_lr)
+
+            result.append(
+                dict(
+                    params=[param],
+                    lr=param_lr,
+                    weight_decay=param_weight_decay,
+                    **{
+                        k: v
+                        for k, v in group.items()
+                        if k not in ("params", "lr", "weight_decay")
+                    },
+                )
+            )
+    return result
+
+
+@inherit_docstring(
+    short_description="An **lr-scaled** version of :class:`torch.optim.SGD` for u-muP."
+)
+class SGD(torch.optim.SGD):
+    def __init__(
+        self,
+        params: ParamsT,
```

Comment:
I'm guessing you can't pass a dictionary of pre-grouped parameters (you can for base torch optimizers)?

Reply:
Ignore this - I see a test further down